lib/unicode_normalize/normalize.rb

```diff
@@ -65,71 +55,156 @@ def self.hangul_decomp_one(target)
     (t==TBASE ? [l, v] : [l, v, t]).pack('U*') + target[1..-1]
   end
 
-  def self.hangul_comp_one(string)
-    length = string.length
-    if length>1 and 0 <= (lead =string[0].ord-LBASE) and lead  < LCOUNT and
-                    0 <= (vowel=string[1].ord-VBASE) and vowel < VCOUNT
-      lead_vowel = SBASE + (lead * VCOUNT + vowel) * TCOUNT
-      if length>2 and 0 < (trail=string[2].ord-TBASE) and trail < TCOUNT
-        (lead_vowel + trail).chr(Encoding::UTF_8) + string[3..-1]
-      else
-        lead_vowel.chr(Encoding::UTF_8) + string[2..-1]
+  def self.hangul_comp_one(a, b)
+    a_ord = a.ord
+    b_ord = b.ord
+
+    # (L, V) -> LV
+    if (LBASE..LLAST).cover?(a_ord) && (VBASE..VLAST).cover?(b_ord)
+      l = a_ord - LBASE
+      v = b_ord - VBASE
+      lv = l * NCOUNT + v * TCOUNT
+      codepoint = SBASE + lv # Codepoint of the composed Hangul
+      return codepoint.chr(Encoding::UTF_8)
+    end
+
+    # (LV, T) -> LVT
+    if (SBASE..SLAST).cover?(a_ord)
+      s = a_ord - SBASE
+      if (s % TCOUNT) == 0 && (TFIRST..TLAST).cover?(b_ord)
+        composed = a_ord + (b_ord - TBASE)
+        return composed.chr(Encoding::UTF_8)
       end
-    else
-      string
     end
+
+    nil
+  end
+
+  def self.to_nfd_arr(string)
+    decomposed_arr = string.each_char.flat_map { decompose_one(it) }
+    ordered_arr = canonical_ordering(decomposed_arr)
+    ordered_arr
   end
 
-  ## Canonical Ordering
-  def self.canonical_ordering_one(string)
-    sorting = string.each_char.collect { |c| [c, CLASS_TABLE[c]] }
-    (sorting.length-2).downto(0) do |i| # almost, but not exactly bubble sort
-      (0..i).each do |j|
-        later_class = sorting[j+1].last
-        if 0<later_class and later_class<sorting[j].last
-          sorting[j], sorting[j+1] = sorting[j+1], sorting[j]
+  def self.to_nfkd_arr(string)
+    kompatibled_arr = string.each_char.flat_map { kompatible_one(it) }
+    decomposed_arr = kompatibled_arr.each.flat_map { decompose_one(it) }
+    ordered_arr = canonical_ordering(decomposed_arr)
```

Comment:
`ordered_arr` is assigned but unused variable.
```suggestion
    canonical_ordering(decomposed_arr)
```